File: Week3/Code/Vectorize1.R
#Author: Ruth Keane (ruth.keane19@imperial.ac.uk)
#Script: Vectorize1.R
#Desc: Summing the elements of a matrix using both a vectorized and non-vectorized function
#Arguments: none
#Input: none
#Output: none
#Date: Oct 2019
#matrix
M <- matrix(runif(1000000),1000,1000)
#sum making a manual function
SumAllElements <- function(M) {
    Dimensions <- dim(M)
    Tot <- 0
    for (i in 1:Dimensions[1]){
        for (j in 1:Dimensions[2]){
            Tot <- Tot + M[i,j]
        }
    }
    return(Tot)
}

print("Using loops, the time taken is:")
print(system.time(SumAllElements(M)))
print("Using the in built vectorized function, the time taken is:")
print(system.time(sum(M)))

File: Week3/Code/Vectorize1.py
#/usr/bin/env python3
#Author: Ruth Keane (ruth.keane19@imperial.ac.uk)
#Script: Vectorize1.py
#Desc: Summing the elements of a matrix using both a vectorized and non-vectorized function
#Arguments: none
#Input: none
#Output: none
#Date: Oct 2019
"""summing matrix elements two different ways and timing this"""
#import packages
from random import seed
from random import random
import numpy as np
import scipy as sc
from timeit import default_timer as timer
#initialise matrix
a=1000
mat=np.random.rand(a,a)
def SumAllElements(x):
    """sum of all elements in a matrix"""
    dim = x.shape
    Tot = 0
    for I in range(0,dim[1]):
        for J in range(0,dim[0]):    
            Tot = Tot + x[J,I]      
    return(Tot)

#time different functions
print("Using loops, the time taken is:")
start=timer()
SumAllElements(mat)
end=timer()
print(end-start)
print("using the in built vectorized function, the time taken is:")
start=timer()
np.sum(mat)
end=timer()
print(end-start)
